crates/uv/tests/it/sync.rs

```diff
@@ -1461,6 +1461,157 @@ fn sync_build_isolation_package() -> Result<()> {
+          ModuleNotFoundError: No module named 'hatchling'
+
+          hint: This error likely indicates that `source-distribution` depends on `hatchling`, but doesn't declare it as a build dependency. If `source-distribution` is a first-party package, consider adding `hatchling` to its `build-system.requires`. Otherwise, either add it to your `pyproject.toml` under:
+
+          [tool.uv.extra-build-dependencies]
+          source-distribution = ["hatchling"]
+
+          or `uv pip install hatchling` into the environment and re-run with `--no-build-isolation`.
+      help: `source-distribution` was included because `project` (v0.1.0) depends on `source-distribution`
+    "#);
+
+    // Add `hatchling`.
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        requires-python = ">=3.12"
+        dependencies = [
+            "hatchling",
+            "source-distribution @ https://files.pythonhosted.org/packages/10/1f/57aa4cce1b1abf6b433106676e15f9fa2c92ed2bd4cf77c3b50a9e9ac773/source_distribution-0.0.1.tar.gz",
+        ]
+        "#,
+    )?;
+
+    // Running `uv sync` should succeed; `hatchling` should be installed first.
+    uv_snapshot!(context.filters(), context.sync().arg("--no-build-isolation-package").arg("source-distribution"), @r"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Resolved 7 packages in [TIME]
+    Prepared 5 isolated packages in [TIME]
+    Installed 5 isolated packages in [TIME]
+    Prepared 1 non-isolated package in [TIME]
+    Installed 1 non-isolated package in [TIME]
+     + hatchling==1.22.4
+     + packaging==24.0
+     + pathspec==0.12.1
+     + pluggy==1.4.0
+     + source-distribution==0.0.1 (from https://files.pythonhosted.org/packages/10/1f/57aa4cce1b1abf6b433106676e15f9fa2c92ed2bd4cf77c3b50a9e9ac773/source_distribution-0.0.1.tar.gz)
+     + trove-classifiers==2024.3.3
+    ");
+
+    // Modify the version.
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        requires-python = ">=3.12"
+        dependencies = [
+            "source-distribution @ https://files.pythonhosted.org/packages/1f/e5/5b016c945d745f8b108e759d428341488a6aee8f51f07c6c4e33498bb91f/source_distribution-0.0.3.tar.gz",
+        ]
+        "#,
+    )?;
+
+    // Running `uv sync` should uninstall `hatchling`, then build `source-distribution`, then uninstall
+    // the existing `source-distribution`, and finally install the new one.
+    uv_snapshot!(context.filters(), context.sync().arg("--no-build-isolation-package").arg("source-distribution"), @r"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Resolved 2 packages in [TIME]
+    Uninstalled 5 isolated packages in [TIME]
+    Prepared 1 non-isolated package in [TIME]
+    Uninstalled 1 non-isolated package in [TIME]
+    Installed 1 non-isolated package in [TIME]
```

Comment:
This was a bit tedious (the labels) but I think it makes a big difference.

Reply:
I have a couple thoughts....

1. "isolated" and "non-isolated" seem a little vague, though I think it's nice it's only present if you're using `no-build-isolation`, you could be running `uv sync` in a project where you did not configure that. I wonder if we should drop "isolated" and say "packages without build isolation" (or similar) for the non-isolated case?
2. I wonder if we can get away with repeating the prefix, e.g., if having it on "Prepared" is sufficient?

```
    Uninstalled 5 packages in [TIME]
    Prepared 1 package without build isolation in [TIME]
    Uninstalled 1 package in [TIME]
    Installed 1 package in [TIME]
```

I guess if we have a cached wheel.. it'd just say "Uninstalled" twice in a row though?

Reply:
Fixed...